GH-35239: [MATLAB] Report error when building the MATLAB Interface to Arrow in Debug mode on Windows (#39595)
apache/arrow: matlab/CMakeLists.txt

  build_arrow()
endif()

# Identify MATLAB build mode (Debug or Release)
if(WIN32)

>   execute_process(COMMAND dumpbin /dependents ${Matlab_MAIN_PROGRAM} OUTPUT_VARIABLE DEPENDENTS)

Reviewer: @kevingurney Is there any MATLAB main program that is built in Debug mode on Windows?

If all MATLAB main programs are built in Release mode on Windows, we don't need to check this.

We just need to check our build mode instead:

```cmake
if(WIN32)
  string(TOUPPER "${CMAKE_BUILD_TYPE}" CMAKE_BUILD_TYPE_UPPER)
  if(CMAKE_BUILD_TYPE_UPPER STREQUAL DEBUG)
    message(FATAL_ERROR "...")
  endif()
endif()
```

Author: Should I resolve the changes? or wait for @kevingurney  confirmation?

Author: > @kevingurney Is there any MATLAB main program that is built in Debug mode on Windows?
> 
> If all MATLAB main programs are built in Release mode on Windows, we don't need to check this.
> 
> We just need to check our build mode instead:
> 
> ```cmake
> if(WIN32)
>   string(TOUPPER "${CMAKE_BUILD_TYPE}" CMAKE_BUILD_TYPE_UPPER)
>   if(CMAKE_BUILD_TYPE_UPPER STREQUAL DEBUG)
>     message(FATAL_ERROR "...")
>   endif()
> endif()
> ```

@kou Should I start working upon this change?

Reviewer: Can we wait for a response from @kevingurney (or @sgilmore10 / @leihou6116 may know)?

Author: Yaa sure :))

Reviewer 2: @kou and @ShaiviAgarwal2 - sorry for delay!

To answer your question: all publicly released versions of MATLAB are built in release mode (i.e. linked against a release version of the MSVC Runtime Libraries).

  if(CMAKE_BUILD_TYPE_UPPER STREQUAL DEBUG)

>     message(FATAL_ERROR "You can't build this with Debug because MATLAB is built with Release. You must use Release")

Reviewer 2: ```suggestion
    message(FATAL_ERROR "Building the MATLAB interface in Debug mode is not supported.")
```
